core/src/main/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerFacade.java

```diff
@@ -38,55 +36,48 @@
     private MediaPlayer.OnErrorListener onErrorListener;
     private MediaPlayer.OnVideoSizeChangedListener onSizeChangedListener;
 
-    private SurfaceHolderRequester surfaceHolderRequester;
+    private MediaPlayerCreator mediaPlayerCreator;
 
     static AndroidMediaPlayerFacade newInstance(Context context) {
         TrackInfosFactory trackInfosFactory = new TrackInfosFactory();
         AndroidMediaPlayerAudioTrackSelector trackSelector = new AndroidMediaPlayerAudioTrackSelector(trackInfosFactory);
         PlaybackStateChecker playbackStateChecker = new PlaybackStateChecker();
         AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
-        return new AndroidMediaPlayerFacade(context, audioManager, trackSelector, playbackStateChecker);
+        MediaPlayerCreator mediaPlayerCreator = new MediaPlayerCreator();
+        return new AndroidMediaPlayerFacade(context, audioManager, trackSelector, playbackStateChecker, mediaPlayerCreator);
     }
 
-    AndroidMediaPlayerFacade(Context context, AudioManager audioManager, AndroidMediaPlayerAudioTrackSelector trackSelector, PlaybackStateChecker playbackStateChecker) {
+    AndroidMediaPlayerFacade(Context context,
+                             AudioManager audioManager,
+                             AndroidMediaPlayerAudioTrackSelector trackSelector,
+                             PlaybackStateChecker playbackStateChecker,
+                             MediaPlayerCreator mediaPlayerCreator) {
         this.context = context;
         this.audioManager = audioManager;
         this.trackSelector = trackSelector;
         this.playbackStateChecker = playbackStateChecker;
-    }
-
-    void setSurfaceHolderRequester(SurfaceHolderRequester surfaceHolderRequester) {
-        this.surfaceHolderRequester = surfaceHolderRequester;
-    }
-
-    void prepareVideo(final Uri videoUri) {
-        if (surfaceHolderRequester == null) {
-            logPlayerNotAttachedWarning("prepareVideo()");
-            return;
+        this.mediaPlayerCreator = mediaPlayerCreator;
+    }
+
+    void prepareVideo(Uri videoUri, SurfaceHolder surfaceHolder) {
+        requestAudioFocus();
+        release();
```

Comment:
potentially this `release` could be remove as we're doing a `reset` before loading the video if the player has previously loaded